funfact/optim/_adam_np.py

```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+import jax.numpy as np
+
+class Adam:
+
+    def __init__(
+        self, x, lr=0.1, beta1=0.9, beta2=0.999, epsilon=1e-7
+    ):
+        self.x = x
+        self.lr = lr
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.epsilon = epsilon
+        self.M = [np.zeros(w.shape) for w in x.factors]
+        self.V = [np.zeros(w.shape) for w in x.factors]
+
+    def step(self, grad):
+        grad_factors = grad.factors
+        new_factors = self.x.factors 
```

Comment:
I'd suggest we keep the details of `Factorization` out of here, i.e., so that `x` is already a list of tensors, and so is `grad`. This should give better decoupling between the modules.

Reply:
After a second thought, I realize that we need to assign to the `.factors` attribute. Let me check how the JAX NN modules handle this situation.